tests/test_qaoa.py

```diff
@@ -199,3 +200,91 @@ def test_maxcut_output(self, graph, target_hamiltonian):
         assert cost_coeffs == target_coeffs
         assert cost_ops == target_ops
         assert cost_wires == target_wires
+
+
+class TestUtils:
+    """Tests that the utility functions are working properly"""
+
+    @pytest.mark.parametrize(
+        ("hamiltonian", "value"),
+        (
+            (qml.Hamiltonian([1, 1], [qml.PauliZ(0), qml.PauliZ(1)]), True),
+            (qml.Hamiltonian([1, 1], [qml.PauliX(0), qml.PauliZ(1)]), False),
+            (qml.Hamiltonian([1, 1], [qml.PauliZ(0) @ qml.Identity(1), qml.PauliZ(1)]), True),
+            (qml.Hamiltonian([1, 1], [qml.PauliZ(0), qml.PauliX(0) @ qml.PauliZ(1)]), False),
+        ),
+    )
+    def test_diagonal_terms(self, hamiltonian, value):
+        assert qaoa.layers._diagonal_terms(hamiltonian) == value
+
+
+class TestLayers:
+    """Tests that the cost and mixer layers are being constructed properly"""
+
+    def test_mixer_layer_errors(self):
+        """Tests that the mixer layer is throwing the correct errors"""
+
+        hamiltonian = [[1, 1], [1, 1]]
+
+        with pytest.raises(ValueError) as info:
+            output = qaoa.mixer_layer(hamiltonian)
+
+        assert "hamiltonian must be of type pennylane.Hamiltonian, got list" in str(info.value)
+
+    def test_cost_layer_errors(self):
+        """Tests that the cost layer is throwing the correct errors"""
+
+        hamiltonian = [[1, 1], [1, 1]]
+
+        with pytest.raises(ValueError) as info:
+            output = qaoa.cost_layer(hamiltonian)
+
+        assert "hamiltonian must be of type pennylane.Hamiltonian, got list" in str(info.value)
+
+        hamiltonian = qml.Hamiltonian([1, 1], [qml.PauliZ(0), qml.PauliX(1)])
+
+        with pytest.raises(ValueError) as info:
+            output = qaoa.cost_layer(hamiltonian)
+
+        assert "hamiltonian must be written only in terms of PauliZ and Identity gates" in str(
+            info.value
+        )
+
+    def test_mixer_layer_output(self):
+        """Tests that the gates of the mixer layer is correct"""
+
+        alpha = 1
+        hamiltonian = qml.Hamiltonian([1, 1], [qml.PauliX(0), qml.PauliX(1)])
```

Comment:
As far as I understand, this is testing that the layers are implementing the same sequence of gates as ApproxTimeEvolution, which I'm not sure is the best way to test since ApproxTimeEvolution may be giving the wrong answers, i.e., these tests don't add much more than the tests for ApproxTimeEvolution. Not sure if this is ok...

Maybe it's good to test an output circuit explicitly, without relying on ApproxTimeEvolution

Reply:
Made these changes 🙂